sympy/core/equation.py

```diff
@@ -0,0 +1,508 @@
+        else:
+            return func(expr, *args, **kwargs)
+
+    def apply(self, func, *args, side='both', **kwargs):
+        """
+        Apply an operation/function/method to the equation returning the
+        resulting equation.
+
+        Parameters
+        ==========
+
+        func: object
+            object to apply usually a function
+
+        side: 'both', 'lhs', 'rhs', optional
+            Specifies which side of the equation the operation will be applied
+            to. Default is 'both'.
+
+         """
+        lhs = self.lhs
+        rhs = self.rhs
+        if side in ('both', 'lhs'):
+            lhs = self._applytoexpr(self.lhs, func, *args, **kwargs)
+        if side in ('both', 'rhs'):
+            rhs = self._applytoexpr(self.rhs, func, *args, **kwargs)
+        return Equation(lhs, rhs, check=False)
+
+    def applylhs(self, func, *args, **kwargs):
+        """
+        If lhs side of the equation has a defined subfunction (attribute) of
+        name ``func``, that will be applied instead of the global function.
+        The operation is applied to only the lhs.
+        """
+        return self.apply(func, *args, **kwargs, side='lhs')
+
+    def applyrhs(self, func, *args, **kwargs):
+        """
+        If rhs side of the equation has a defined subfunction (attribute) of
+        name ``func``, that will be applied instead of the global function.
+        The operation is applied to only the rhs.
+        """
+        return self.apply(func, *args, **kwargs, side='rhs')
+
+    class _Sides:
+        def __init__(self,eqn, side='both'):
+            self.eqn = eqn
+            self.side = side
+
+        def __getattr__(self, name):
+            func = None
+            try:
+                func = getattr(self.eqn.rhs, name)
+            except AttributeError:
+                func = getattr(self.eqn.lhs, name)
+            return functools.partial(self.eqn.apply, func, side=self.side)
+
+    @property
+    def do(self):
+        return self._Sides(self,side='both')
+
+    @property
+    def dolhs(self):
+        return self._Sides(self,side='lhs')
+
+    @property
+    def dorhs(self):
+        return self._Sides(self,side='rhs')
+
+    #####
+    # Overrides of binary math operations
+    #####
+
+    @classmethod
+    def _binary_op(cls, a, b, opfunc_ab):
+        if isinstance(a, Equation) and not isinstance(b, Equation):
+            return Equation(opfunc_ab(a.lhs, b), opfunc_ab(a.rhs, b),
+                            check=False)
+        elif isinstance(b, Equation) and not isinstance(a, Equation):
+            return Equation(opfunc_ab(a, b.lhs), opfunc_ab(a, b.rhs),
+                            check=False)
+        elif isinstance(a, Equation) and isinstance(b, Equation):
+            return Equation(opfunc_ab(a.lhs, b.lhs), opfunc_ab(a.rhs, b.rhs),
+                            check=False)
+        else:
+            raise TypeError('One of a or b should be an equation')
```

Comment:
This should probably be `return NotImplemented`

Reply:
changed in commit 6419fdad